pep-0692.rst

```diff
@@ -0,0 +1,438 @@
+        animal: Dog
+        string: str
+        number: NotRequired[int]
+
+    def src(animal: Animal, string: str, number: int = ...): ...
+    def dest(**kwargs: **Example): ...
+
+    dest = src  # OK!
+
+4. The destination callable contains ``**kwargs: **TypedDict`` and the source
+callable contains untyped ``**kwargs``:
+
+::
+
+    def src(**kwargs): ...
+    def dest(**kwargs: **Movie): ...
+
+    dest = src  # OK!
+
+5. The destination callable contains ``**kwargs: **TypedDict``, the source
+callable contains traditionally typed ``**kwargs: T`` and each of the
+destination function ``TypedDict``\'s fields is assignable to a variable of type
+``T``.
+
+On the other hand, if the destination callable contains either untyped or
+traditionally typed ``**kwargs: T`` and the source callable is typed using
+``**kwargs: **TypedDict`` then an error should be generated, because
+traditionally typed ``**kwargs`` aren't checked for keyword names.
+
+To sum up points, functions' parameters should behave contravariantly. In
+addition, functions' return types should behave covariantly. 
+
+Grammar Changes
+===============
+This PEP requires a grammar change so that the double asterisk syntax is allowed
+for ``**kwargs`` annotations. The proposed change is to extend the ``kwds`` rule
+as follows:
+
+Before:
+
+::
+
+    kwds: '**' param_no_default 
+
+After:
+
+::
+
+    kwds:
+        | '**' param_no_default_double_star_annotation
+        | '**' param_no_default
+
+    param_no_default_double_star_annotation:
+        | param_double_star_annotation & ')'
+
+    param_double_star_annotation: NAME double_star_annotation
+
+    double_star_annotation: ':' double_star_expression
+
+    double_star_expression: '**' expression
+
+A new AST node needs to be created so that type checkers can differentiate the
+semantics of the new syntax and take into account that the ``**kwargs`` should
+be unpacked. The proposition is to add a new ``DoubleStarred`` AST node. Then,
+an AST node for the function defined as:
+
+::
+
+    def foo(**kwargs: **Movie): ...
+
+should look as below:
+
+::
+
+    FunctionDef(
+      name='foo',
+      args=arguments(
+        posonlyargs=[],
+        args=[],
+        kwonlyargs=[],
+        kw_defaults=[],
+        kwarg=arg(
+          arg='kwargs',
+          annotation=DoubleStarred(
+            value=Name(id='Movie', ctx=Load()),
+            ctx=Load())),
+        defaults=[]),
+      body=[
+        Expr(
+          value=Constant(value=Ellipsis))],
+      decorator_list=[])],
+    type_ignores=[])
+
+The runtime annotations should be consistent with the AST. Continuing the
+previous example:
+
+::
+
+    >>> def foo(**kwargs: **Movie): ...
+    ...
+    >>> foo.__annotations__
+    {'kwargs': **Movie}
+
+The double asterisk operator should call the ``__unpack__`` special method on
+the object it was used on. This means that ``def foo(**kwargs: **T): ...`` is
+equivalent to ``def foo(**kwargs: T.__unpack__()): ...``.
```

Comment:
This appears to imply that an entirely new `**expr` operator (with related `__unpack__` dunder) is being introduced, that could potentially be used _anywhere_. I would advocate narrowing the scope of the PEP to only allow the new `**expr` syntax in the position of a kwarg annotation and not (yet) elsewhere else.

Reply:
Based on reading other comments it sounds like there is _not_ an intent to introduce a new general operator. In that case I might suggest replacing the word "operator" here with some other word like "syntax":

```suggestion
The double asterisk syntax should call the ``__unpack__`` special method on
the object it was used on. This means that ``def foo(**kwargs: **T): ...`` is
equivalent to ``def foo(**kwargs: T.__unpack__()): ...``.
```